File: implementations/java/server/src/test/java/com/theshuai/specusserver/management/service/HttpMediaManifestSupportTests.java
package com.theshuai.specusserver.management.service;

import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

class HttpMediaManifestSupportTests {

    @Test
    void classifiesRangeAndOpaqueMediaResponsesOnEnabledMediaRoutes() {
        assertThat(HttpMediaManifestSupport.classify(
                "/stream?id=42", "video/mp4", 206, "bytes 0-99/1000"))
                .isEqualTo(HttpMediaManifestSupport.PROGRESSIVE);
        assertThat(HttpMediaManifestSupport.classify(
                "/segment?id=42", "application/octet-stream", 200, null))
                .isEqualTo(HttpMediaManifestSupport.MEDIA_SEGMENT);
        assertThat(HttpMediaManifestSupport.classify(
                "/keys/session.key", "application/binary", 200, null))
                .isEqualTo(HttpMediaManifestSupport.MEDIA_SEGMENT);
    }

    @Test
    void parsesAndRewritesHlsReferences() {
        String manifest = """
                #EXTM3U
                #EXT-X-MEDIA-SEQUENCE:12
                #EXT-X-MAP:URI="init.mp4"
                #EXT-X-KEY:METHOD=AES-128,URI="secret.key"
                #EXTINF:4,
                part-12.m4s
                """;

        HttpMediaManifestSupport.ParsedManifest parsed = HttpMediaManifestSupport.parse(
                HttpMediaManifestSupport.HLS_MANIFEST, "/media/live/index.m3u8", manifest);

        assertThat(parsed.live()).isTrue();
        assertThat(parsed.references())
                .extracting(HttpMediaManifestSupport.ManifestReference::relationType)
                .containsExactly("INITIALIZATION", "KEY", "SEGMENT");
        assertThat(parsed.references().get(2).sequence()).isEqualTo(12L);
        assertThat(parsed.references().get(2).resolvedSourceUrl()).isEqualTo("/media/live/part-12.m4s");

        String rewritten = HttpMediaManifestSupport.rewrite(
                HttpMediaManifestSupport.HLS_MANIFEST,
                "/media/live/index.m3u8",
                manifest,
                "/api/public/media-playback/ticket/asset");
        assertThat(rewritten)
                .contains("url=%2Fmedia%2Flive%2Finit.mp4")
                .contains("url=%2Fmedia%2Flive%2Fsecret.key")
                .contains("url=%2Fmedia%2Flive%2Fpart-12.m4s");
    }

    @Test
    void keepsDashTemplateTokensAvailableForThePlayer() {
        String manifest = """
                <MPD type="dynamic">
                  <Period><AdaptationSet>
                    <SegmentTemplate initialization="init-$RepresentationID$.mp4"
                                     media="chunk-$Number$.m4s"/>
                  </AdaptationSet></Period>
                </MPD>
                """;

        String rewritten = HttpMediaManifestSupport.rewrite(
                HttpMediaManifestSupport.DASH_MANIFEST,
                "/dash/manifest.mpd",
                manifest,
                "/api/public/media-playback/ticket/asset");

        assertThat(rewritten)
                .contains("$RepresentationID$")
                .contains("$Number$")
                .contains("url=%2Fdash%2Fchunk-$Number$.m4s");
    }

    @Test
    void rejectsInvalidContentRanges() {
        assertThat(HttpMediaManifestSupport.parseContentRange("bytes 10-19/20"))
                .isEqualTo(new HttpMediaManifestSupport.ContentRange(10, 19, 20L));
        assertThat(HttpMediaManifestSupport.parseContentRange("bytes 20-10/30")).isNull();
        assertThat(HttpMediaManifestSupport.parseContentRange("bytes 0-30/30")).isNull();
    }
}

File: implementations/csharp/server/tests/Specus.IntegrationTests/HttpMediaManifestSupportTests.cs
using Specus.Server.Management;

namespace Specus.IntegrationTests;

public sealed class HttpMediaManifestSupportTests
{
    [Fact]
    public void ClassifiesRangeOpaqueAndManifestResponsesLikeJava()
    {
        Assert.Equal(HttpMediaManifestSupport.Progressive,
            HttpMediaManifestSupport.Classify("/stream?id=42", "video/mp4", 206,
                "bytes 0-99/1000"));
        Assert.Equal(HttpMediaManifestSupport.MediaSegment,
            HttpMediaManifestSupport.Classify("/segment?id=42", "application/octet-stream", 200,
                null));
        Assert.Equal(HttpMediaManifestSupport.MediaSegment,
            HttpMediaManifestSupport.Classify("/keys/session.key", "application/binary", 200,
                null));
        Assert.Equal(HttpMediaManifestSupport.HlsManifest,
            HttpMediaManifestSupport.Classify("/index.m3u8", "text/plain", 200, null));
    }

    [Fact]
    public void ParsesAndRewritesHlsReferences()
    {
        const string manifest = """
            #EXTM3U
            #EXT-X-MEDIA-SEQUENCE:12
            #EXT-X-MAP:URI="init.mp4"
            #EXT-X-KEY:METHOD=AES-128,URI="secret.key"
            #EXTINF:4,
            part-12.m4s
            """;

        var parsed = HttpMediaManifestSupport.Parse(HttpMediaManifestSupport.HlsManifest,
            "/media/live/index.m3u8", manifest);

        Assert.True(parsed.Live);
        Assert.Equal(["INITIALIZATION", "KEY", "SEGMENT"],
            parsed.References.Select(reference => reference.RelationType));
        Assert.Equal(12, parsed.References[2].Sequence);
        Assert.Equal("/media/live/part-12.m4s", parsed.References[2].ResolvedSourceUrl);
        var rewritten = HttpMediaManifestSupport.Rewrite(HttpMediaManifestSupport.HlsManifest,
            "/media/live/index.m3u8", manifest,
            "/api/public/media-playback/ticket/asset");
        Assert.Contains("url=%2Fmedia%2Flive%2Finit.mp4", rewritten,
            StringComparison.OrdinalIgnoreCase);
        Assert.Contains("url=%2Fmedia%2Flive%2Fsecret.key", rewritten,
            StringComparison.OrdinalIgnoreCase);
        Assert.Contains("url=%2Fmedia%2Flive%2Fpart-12.m4s", rewritten,
            StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void DashRewritePreservesPlayerTemplateTokens()
    {
        const string manifest = """
            <MPD type="dynamic">
              <Period><AdaptationSet>
                <SegmentTemplate initialization="init-$RepresentationID$.mp4"
                                 media="chunk-$Number$.m4s"/>
              </AdaptationSet></Period>
            </MPD>
            """;

        var rewritten = HttpMediaManifestSupport.Rewrite(HttpMediaManifestSupport.DashManifest,
            "/dash/manifest.mpd", manifest,
            "/api/public/media-playback/ticket/asset");

        Assert.Contains("$RepresentationID$", rewritten, StringComparison.Ordinal);
        Assert.Contains("$Number$", rewritten, StringComparison.Ordinal);
        Assert.Contains("url=%2Fdash%2Fchunk-$Number$.m4s", rewritten,
            StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void RejectsInvalidContentRanges()
    {
        Assert.Equal(new HttpMediaManifestSupport.ContentRange(10, 19, 20),
            HttpMediaManifestSupport.ParseContentRange("bytes 10-19/20"));
        Assert.Null(HttpMediaManifestSupport.ParseContentRange("bytes 20-10/30"));
        Assert.Null(HttpMediaManifestSupport.ParseContentRange("bytes 0-30/30"));
    }
}
